<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>
  Class: Blast
  
    &mdash; Documentation by YARD 0.9.26
  
</title>

  <link rel="stylesheet" href="css/style.css" type="text/css" />

  <link rel="stylesheet" href="css/common.css" type="text/css" />

<script type="text/javascript">
  pathId = "Blast";
  relpath = '';
</script>


  <script type="text/javascript" charset="utf-8" src="js/jquery.js"></script>

  <script type="text/javascript" charset="utf-8" src="js/app.js"></script>


  </head>
  <body>
    <div class="nav_wrap">
      <iframe id="nav" src="class_list.html?1"></iframe>
      <div id="resizer"></div>
    </div>

    <div id="main" tabindex="-1">
      <div id="header">
        <div id="menu">
  
    <a href="_index.html">Index (B)</a> &raquo;
    
    
    <span class="title">Blast</span>
  
</div>

        <div id="search">
  
    <a class="full_list_link" id="class_list_link"
        href="class_list.html">

        <svg width="24" height="24">
          <rect x="0" y="4" width="24" height="4" rx="1" ry="1"></rect>
          <rect x="0" y="12" width="24" height="4" rx="1" ry="1"></rect>
          <rect x="0" y="20" width="24" height="4" rx="1" ry="1"></rect>
        </svg>
    </a>
  
</div>
        <div class="clear"></div>
      </div>

      <div id="content"><h1>Class: Blast
  
  
  
</h1>
<div class="box_info">
  
  <dl>
    <dt>Inherits:</dt>
    <dd>
      <span class="inheritName">Object</span>
      
        <ul class="fullTree">
          <li>Object</li>
          
            <li class="next">Blast</li>
          
        </ul>
        <a href="#" class="inheritanceTree">show all</a>
      
    </dd>
  </dl>
  

  
  
  
  
  

  

  
  <dl>
    <dt>Defined in:</dt>
    <dd>Class_blast.rb</dd>
  </dl>
  
</div>








  
    <h2>
      Class Method Summary
      <small><a href="#" class="summary_toggle">collapse</a></small>
    </h2>

    <ul class="summary">
      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="#createdb-class_method" title="createdb (class method)">.<strong>createdb</strong>(seqfile)  &#x21d2; Object </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Before start we have to create the database, to do that I search this method in <a href="https://sequenceserver.com/doc">sequenceserver.com/doc</a>/.</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="#make_blast-class_method" title="make_blast (class method)">.<strong>make_blast</strong>(type, database, query)  &#x21d2; Object </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'></div></span>
  
</li>

      
    </ul>
  



  <div id="class_method_details" class="method_details_list">
    <h2>Class Method Details</h2>

    
      <div class="method_details first">
  <h3 class="signature first" id="createdb-class_method">
  
    .<strong>createdb</strong>(seqfile)  &#x21d2; <tt>Object</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Before start we have to create the database, to do that I search this method in <a href="https://sequenceserver.com/doc">sequenceserver.com/doc</a>/</p>


  </div>
</div>
<div class="tags">
  

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


8
9
10
11
12
13
14
15
16
17
18</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'Class_blast.rb', line 8</span>

<span class='kw'>def</span> <span class='kw'>self</span><span class='period'>.</span><span class='id identifier rubyid_createdb'>createdb</span><span class='lparen'>(</span><span class='id identifier rubyid_seqfile'>seqfile</span><span class='rparen'>)</span> <span class='comment'># Before start we have to create the database, to do that I search this method in https://sequenceserver.com/doc/
</span>  <span class='id identifier rubyid_system'>system</span> <span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>mkdir Database</span><span class='tstring_end'>&#39;</span></span>
<span class='comment'>#to create the database I have to guess if the seqs are of a protein or a nucleotide,to do that I used the information on http://bioruby.org/rdoc/Bio/Sequence.html 
</span>  <span class='id identifier rubyid_sequence'>sequence</span><span class='op'>=</span> <span class='const'>Bio</span><span class='op'>::</span><span class='const'>Sequence</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='const'>Bio</span><span class='op'>::</span><span class='const'>FlatFile</span><span class='period'>.</span><span class='id identifier rubyid_auto'>auto</span><span class='lparen'>(</span><span class='id identifier rubyid_seqfile'>seqfile</span><span class='rparen'>)</span><span class='period'>.</span><span class='id identifier rubyid_next_entry'>next_entry</span><span class='period'>.</span><span class='id identifier rubyid_seq'>seq</span><span class='rparen'>)</span>
  <span class='kw'>if</span> <span class='id identifier rubyid_sequence'>sequence</span><span class='period'>.</span><span class='id identifier rubyid_guess'>guess</span> <span class='op'>==</span> <span class='const'>Bio</span><span class='op'>::</span><span class='const'>Sequence</span><span class='op'>::</span><span class='const'>AA</span>
    <span class='id identifier rubyid_system'>system</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>cp </span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_seqfile'>seqfile</span><span class='embexpr_end'>}</span><span class='tstring_content'> ./Database|cd Database|makeblastdb -dbtype prot  -title &#39;Target_db&#39;  -parse_seqids -in Database/</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_seqfile'>seqfile</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span>
  <span class='kw'>else</span>
    <span class='id identifier rubyid_system'>system</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>cp </span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_seqfile'>seqfile</span><span class='embexpr_end'>}</span><span class='tstring_content'> ./Database|cd Database|makeblastdb -dbtype nucl  -title &#39;Target_db&#39;  -parse_seqids -in Database/</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_seqfile'>seqfile</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span>

  <span class='kw'>end</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      <div class="method_details ">
  <h3 class="signature " id="make_blast-class_method">
  
    .<strong>make_blast</strong>(type, database, query)  &#x21d2; <tt>Object</tt> 
  

  

  
</h3><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'Class_blast.rb', line 20</span>

<span class='kw'>def</span> <span class='kw'>self</span><span class='period'>.</span><span class='id identifier rubyid_make_blast'>make_blast</span><span class='lparen'>(</span><span class='id identifier rubyid_type'>type</span><span class='comma'>,</span> <span class='id identifier rubyid_database'>database</span><span class='comma'>,</span> <span class='id identifier rubyid_query'>query</span><span class='rparen'>)</span>
     <span class='id identifier rubyid_vhits'>vhits</span><span class='op'>=</span> <span class='lbracket'>[</span><span class='rbracket'>]</span> <span class='comment'>#in this vector I will put the hits that meets the requirements
</span>     <span class='id identifier rubyid_blast'>blast</span><span class='op'>=</span><span class='const'>Bio</span><span class='op'>::</span><span class='const'>Blast</span><span class='period'>.</span><span class='id identifier rubyid_local'>local</span><span class='lparen'>(</span><span class='id identifier rubyid_type'>type</span><span class='comma'>,</span><span class='id identifier rubyid_database'>database</span><span class='rparen'>)</span>
     <span class='kw'>for</span> <span class='id identifier rubyid_seq'>seq</span> <span class='kw'>in</span> <span class='const'>Bio</span><span class='op'>::</span><span class='const'>FlatFile</span><span class='period'>.</span><span class='id identifier rubyid_auto'>auto</span><span class='lparen'>(</span><span class='id identifier rubyid_query'>query</span><span class='rparen'>)</span>
       <span class='id identifier rubyid_report'>report</span> <span class='op'>=</span> <span class='id identifier rubyid_blast'>blast</span><span class='period'>.</span><span class='id identifier rubyid_query'>query</span><span class='lparen'>(</span><span class='id identifier rubyid_seq'>seq</span><span class='rparen'>)</span>
       <span class='id identifier rubyid_report'>report</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_hit'>hit</span><span class='op'>|</span>
         <span class='kw'>if</span> <span class='id identifier rubyid_hit'>hit</span> <span class='kw'>and</span> <span class='id identifier rubyid_hit'>hit</span><span class='period'>.</span><span class='id identifier rubyid_evalue'>evalue</span> 
<span class='comment'>#The params used for filter the hits are from this website https://www.metagenomics.wiki/tools/blast/evalue 
</span><span class='comment'>#the e value threshol is from this webpage as the e value threshold is for huge cuality hits, I use a bigger value,
</span><span class='comment'># in other website I see a good hit has e value from 1e-10 and 1e-50 so I selected 1e-30. The link is https://resources.qiagenbioinformatics.com/manuals/clcgenomicsworkbench/650/_E_value.html#:~:text=The%20default%20threshold%20for%20the,most%20likely%20generate%20more%20hits.
</span><span class='comment'>#the bit score is from https://www.unmc.edu/bsbc/education/courses/gcba815/Week3_Blast.pdf
</span>           <span class='kw'>if</span> <span class='id identifier rubyid_hit'>hit</span><span class='period'>.</span><span class='id identifier rubyid_evalue'>evalue</span> <span class='op'>&lt;=</span> <span class='float'>1e-30</span> <span class='kw'>and</span> <span class='id identifier rubyid_hit'>hit</span><span class='period'>.</span><span class='id identifier rubyid_bit_score'>bit_score</span><span class='op'>&gt;</span><span class='int'>40</span> 
             <span class='id identifier rubyid_vhits'>vhits</span><span class='period'>.</span><span class='id identifier rubyid_append'>append</span><span class='lparen'>(</span><span class='lbracket'>[</span><span class='id identifier rubyid_seq'>seq</span><span class='period'>.</span><span class='id identifier rubyid_entry_id'>entry_id</span><span class='comma'>,</span> <span class='id identifier rubyid_hit'>hit</span><span class='period'>.</span><span class='id identifier rubyid_target_id'>target_id</span><span class='rbracket'>]</span><span class='rparen'>)</span>
             
           <span class='kw'>end</span>
         <span class='kw'>end</span>
      
       <span class='kw'>end</span> 
    
    <span class='kw'>end</span>
    <span class='kw'>return</span> <span class='id identifier rubyid_vhits'>vhits</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
  </div>

</div>

      <div id="footer">
  Generated on Mon Jan  2 23:02:42 2023 by
  <a href="http://yardoc.org" title="Yay! A Ruby Documentation Tool" target="_parent">yard</a>
  0.9.26 (ruby-3.0.0).
</div>

    </div>
  </body>
</html>